cmake/Caffe2Config.cmake.in

```diff
@@ -86,11 +86,7 @@ if(@USE_CUDA@)
   set(CAFFE2_USE_CUDA @USE_CUDA@)
   set(CAFFE2_USE_TENSORRT @USE_TENSORRT@)
 
-  # Add current directory to module path so we pick up FindCUDAToolkit.cmake
-  set(old_CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}")
-  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}")
```

Comment:
This is wrong, the module path is different when torch is installed.

Reply:
We always use the same module as building, that is cmake/module/findcudatoolkit.cmake.  Because cuda.cmake has list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/../Modules_CUDA_fix) in line 27

Reply:
I moved the code of  cmake_module_path rewriting into cuda.cmake

Reply:
`FindCUDAToolkit.cmake` is not installed under `Module_CUDA_fix`...

Reply:
I also used list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/../Modules) in line 28

Reply:
So the modules from the two directories are searched before other places

Reply:
You are wrong, the install path is shown in the build log:
```
C:\Jenkins\Miniconda3\Lib\site-packages\torch\share\cmake\Caffe2\FindCUDAToolkit.cmake
```

Reply:
@peterbell10 thank you for the thorough review. Do you know if we have any tests for `libtorch` usage in CI?  (I know there are bunch of extensions tests, but probably not libtorch ones)

But on the other hand, why wants to know anything about CUDA when linking against libtorch?

Reply:
> Do you know if we have any tests for libtorch usage in CI? 

Yes there is `test/custom_backend` and `test/custom_operator` which use `find_package(Torch)` to build executables.

> But on the other hand, why wants to know anything about CUDA when linking against libtorch?

I believe CMake needs to have all linked targets available, even for private dependencies. The reason, quoting from the [cmake mailing list](https://cmake.org/pipermail/cmake/2016-May/063400.html):

> Static libraries do not carry information about other libraries they depend on. For this reason, when A links in B as PRIVATE and another target C links in A, CMake will still add B to the list of libraries to be linked for C because parts of B are needed by A, but A itself doesn't have that dependency encoded into it. So even though B is an internal implementation detail of A, C still needs B added to the linker command, which CMake conveniently handles for you.

Perhaps there is a way to work around this for shared libraries though.

Reply:
CMake needn't find private libraries again in exposed caffe2.cmake and torch.cmake in principle . Of course, the user needs to ensure the so files are in PATH.  I would like to make dependencies private as possible and remove finding code from public config files. It not only simplifies the code and avoids many unnecessary errors in rare environments such as force_cpu

Reply:
> You are wrong, the install path is shown in the build log:
> 
> ```
> C:\Jenkins\Miniconda3\Lib\site-packages\torch\share\cmake\Caffe2\FindCUDAToolkit.cmake
> ```

Thank you for pointing it out. May I put FindCUDAToolkit.cmake in torch\share\cmake\Caffe2\public so that it is in the same location with other modules and more importantly cuda.cmake will find it ?

Reply:
Then you would need to add `${CMAKE_CURRENT_LIST_DIR}/public` to the module path. What's wrong with leaving it how it was?

Reply:
> Your idea of protecting CMAKE_MODULE_PATH is great, I want to make sure it is also protected when building 

I strongly disagree. During build we want any submodules to pick up the same Find modules that we supply to avoid discrepancies between build configs. But when included as part of `find_package(Torch)` it would be incredibly rude to override `CMAKE_MODULE_PATH` so IMO these shouldn't be treated the same.

Reply:
This PR makes sure that CMAKE_MODULE_PATH  is changed temporarily to prefer our variant of  FindCUDAToolkits.cmake in building pytorch and using **find_package(Torch)** from other projects. And the master version makes sure that CMAKE_MODULE_PATH  is changed temporarily by **find_package(Torch)** and _permanently_ when building pytorch. So the only difference in behavior is an additional protection in building pytorch.

Reply:
@cyyever let me repeat:
> During build we want any submodules to pick up the same Find modules that we supply to avoid discrepancies between build configs.

i.e. we want it changed permanently when building pytorch.